### PR TITLE
please, look at comment

### PR DESCRIPTION
**Pull request by prajdziswet**

### CODE REVIEW

File: ClassLibrary/ClassLibrary/CardReader.cs [hunk 1]
```diff
@@ -1,12 +1,70 @@
...
+        public CardReader(String NameReader, String LastNameReader)
+        {
+
+            if (NameReader.IsNullOrWhiteSpace())
+            {
+                throw new ArgumentNullException("Name is empty");
+            }
+            else if (LastNameReader.IsNullOrWhiteSpace())
+            {
+                throw new ArgumentNullException("Name is empty");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/CardReader.cs]: exception message is incorrect

File: ClassLibrary/ClassLibrary/Library.cs [hunk 2]
```diff
@@ -15,17 +15,16 @@ public List<Book> Books
 
         public void AddBook(String ISBN, String nameBook, Author author)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: Please use Book as a parameter

File: ClassLibrary/ClassLibrary/Library.cs [hunk 3]
```diff
@@ -43,11 +42,6 @@ public List<Author> Authors
 
         public Author AddAuthor(String nameAuthor,String lastNameAuthor)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: Please use Author as a parameter

File: ClassLibrary/ClassLibrary/Library.cs [hunk 4]
```diff
@@ -60,6 +54,27 @@ public Author AddAuthor(String nameAuthor,String lastNameAuthor)
...
 
+        public List<CardReader> CardReaders
+        {
+            get;
+            private set;
+        }
+            = new List<CardReader>();
+
+        //I'm breaking the rule AV1115, but "AddAuthor" breaking the rule too
+        public CardReader GetCardForReader(String NameReader,String LastNameReader)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: there are two very different cases - Getting a reader and Searching for a reader.
you can search by name and last name, but to Get one reader you should have reader's ID.

Usually system user Searches for entity by non-unique parameters like name and then in search results finds one - and then ID of that entity is user in next steps.

bottom line: in current situation, when we don't have users, we don't need Search, only Get - meaning you should use only ID as a parameter.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary/Library.cs (hunk 4 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: While I delete this method. ==
- I did instead of him AddReader.
- While not realize test..

File: ClassLibrary/ClassLibrary.Test/TestAuthors.cs [hunk 5]
```diff
@@ -81,22 +66,16 @@ public void AddNewAuthors()
             Assert.IsTrue(lib.Authors[2].LastName == "Chehov");
         }
 
-        [Test]
-        public void AddSimularAuthors()
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestAuthors.cs:85]: Why did you remove this test case?

File: ClassLibrary/ClassLibrary/Library.cs [hunk 6]
```diff
@@ -13,24 +13,28 @@ public List<Book> Books
         }
             = new List<Book>();
 
-        public void AddBook(String ISBN, String nameBook, Author author)
+        public void AddBookInLibrary(Book book)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: `InLibrary` is not necessary, code `library.AddBook()` is clear enough.

File: ClassLibrary/ClassLibrary/Library.cs [hunk 7]
```diff
@@ -13,24 +13,28 @@ public List<Book> Books
...
         {
-            if (ISBN.IsNullOrWhiteSpace() || nameBook.IsNullOrWhiteSpace() || author == null)
+            if (book == null)
             {
-                throw new ArgumentNullException("One of the fields of the book is empty");
+                throw new NullReferenceException("Book's reference is null");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: Rule of thumb: you never throw NullReferenceException (NRE) yourself. ArgumentNullException is the one programmer uses, NRE is just for runtime.

File: ClassLibrary/ClassLibrary/Library.cs [hunk 8]
```diff
@@ -41,23 +45,51 @@ public List<Author> Authors
         } 
             = new List<Author>();
 
-        public Author AddAuthor(String nameAuthor,String lastNameAuthor)
+        public void AddAuthorInList(Author author)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: `InList` is not necessary here, because calling code should not care if author is added to a list or anywhere else - it just should stay in the library.

File: ClassLibrary/ClassLibrary/Library.cs [hunk 9]
```diff
@@ -41,23 +45,51 @@ public List<Author> Authors
...
 
             if (returnAuthor == null)
             {
-                returnAuthor = new Author(nameAuthor, lastNameAuthor);
-                Authors.Add(returnAuthor);
+                Authors.Add(author);
+            }
+            else
+            {
+                throw new ArgumentException("This author exists in List");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: better to say `$"Author [{author.ID}] {author.Name} {author.LastName} already added to the library"`

File: ClassLibrary/ClassLibrary/Library.cs [hunk 10]
```diff
@@ -41,23 +45,51 @@ public List<Author> Authors
...
-            if (nameAuthor.IsNullOrWhiteSpace() || lastNameAuthor.IsNullOrWhiteSpace())
+            if (author == null)
             {
-                throw new NullReferenceException("The author's first or last name is empty ");
+                throw new NullReferenceException("Author's reference is null");
             }
 
-            Author returnAuthor = Authors?.FirstOrDefault(author =>
-                author.Name == nameAuthor && author.LastName == lastNameAuthor);
+            Author returnAuthor = Authors?.FirstOrDefault(x =>
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: `returnAuthor` is not the best name for this variable.

File: ClassLibrary/ClassLibrary/Library.cs [hunk 11]
```diff
@@ -41,23 +45,51 @@ public List<Author> Authors
...
+            private set;
+        }
+            = new List<Reader>();
+
+
+        public void AddReader(Reader reader)
+        {
+            if (reader == null)
+            {
+                throw new NullReferenceException("Reader reference is null");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: Don't use NRE.

File: ClassLibrary/ClassLibrary/Reader.cs [hunk 12]
```diff
@@ -1,11 +1,70 @@
...
+            get;
+            private set;
+        }
+        public String LastName
+        {
+            get;
+            private set;
+        }
+
+        public List<Book> Books
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Reader.cs]: `Books` is not specific enough, try `BorrowedBooks`.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary/Library.cs (hunk 7 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: не совсем понятно, если во-время выполнения. зачем давать возможность его выбрать?.
Для меня они равноценны, даже NRE - более описвает ситуацию (AV1205):
- да book, author -это аргументы и задать null как таковой у нас нет возможности = потому что конструкторы не позволяют задать нулл полям
- book, author, как классы, передаються по ссылке (а не по значению)? => тоесть нам поступаем NullReference а не агрумент, как таковой

или я что то путаю? 
мне кажеться комент не совсем раскрывает смысл, то что ты хотел сказать (по мне они "почти" равноценны)

File: ClassLibrary/ClassLibrary/Library.cs (hunk 10 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: What idea do you have?

File: ClassLibrary/ClassLibrary/Library.cs (hunk 11 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: while I close this conversation, it's on top.

File: ClassLibrary/ClassLibrary/Reader.cs (hunk 12 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Reader.cs]: while I close this conversation.
I didn't catch that, and "books" - it's all books/
 BorrowedBook exsist in Readers...

File: ClassLibrary/ClassLibrary/Library.cs (hunk 7 above)
Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary/Library.cs]: NRE is thrown when a reference that shouldn't be null was unexpectedly null. It happens automatically, like OutOfMemory or StackOverflow exceptions. when you throw it in a different situation it just a confusion for calling code and someone who will investigate the reason of that exception.

in scope of AV1205 - NRE describes **another** situation, so it's not more specific, it's just incorrect. ArgumentNullException is a correct one for this situation, when an argument that should not be null still is.

File: ClassLibrary/ClassLibrary/Library.cs (hunk 10 above)
Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary/Library.cs]: try to think about it yourself, it's also ok to leave it as is.

File: ClassLibrary/ClassLibrary/Reader.cs (hunk 12 above)
Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary/Reader.cs]: now I'm confused.. isn't this Reader class?

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary/Reader.cs (hunk 12 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Reader.cs]: тьфу, I messed up.... I thought about the library 
ps I agree with you...

File: ClassLibrary/ClassLibrary/Library.cs (hunk 7 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: ок, thaks - I understant

File: ClassLibrary/ClassLibrary.Test/TestEntities.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        public void DifferentIDReaders()
+        {
+            Reader reader = new Reader("Lev", "Tolstoj");
+            Reader reader1 = new Reader("Alexsandr", "Tolstoj");
+
+            reader.ID.ShouldNotBe(reader1.ID);
+        }
+
+        [Test]
+        public void AddNullBookInCard()
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestEntities.cs:108]: Positive case for `AddBookInCard` is missing

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        public void FindBookNameBook()
+        {
+            RepositoryBooks RB = new RepositoryBooks();
+            Author author = new Author("Lev", "Tolstoj");
+            Book book = new Book("226611156", "War and Peace", author);
+            RB.AddBook(book);
+
+            List<Book> books = RB.FindBooks("War and Peace");
+
+            books[0].ShouldBe(book);
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]: If there should be only one book in the response, then check that too.

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+        [Test]
+        public void FindBookLastNameAuthor()
+        {
+            RepositoryBooks RB = new RepositoryBooks();
+            Author author = new Author("Lev", "Tolstoj");
+            Book book = new Book("226611156", "War and Peace", author);
+            RB.AddBook(book);
+
+            List<Book> books = RB.FindBooks("", "", "Tolstoj");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]: here you can see how default parameters make code even difficult to write - you have to specify two empty parameters that are not needed. usually, DTO class is created for Find requests.

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs [hunk 16]
```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Collections.Generic;
+using NUnit.Framework;
+using Shouldly;
+
+namespace ClassLibrary.Test
+{
+    [TestFixture]
+    public class TestRepositoryBooks
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs:10]: You probably already noticed that testing the `Find` method is difficult. Actually, it's even more difficult :) There are 3 parameters, each of the parameters can be omitted, meaning that to cover only positive searches you need 8 test cases. And the same amount of cases is needed for negative cases (well, except for the "all parameters empty" case, which is always positive).

Also, for multiple search results you can use only one test case, "return 2 of 3", but for the full coverage, it'll be also 8 cases (7, actually, because of the "all parameters empty" case).

that's a lot, so please look into [TestCase feature](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcase.html)

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+            Author author = new Author("Lev", "Tolstoj");
+            Book book = new Book("226611156", "War and Peace", author);
+            RB.AddBook(book);
+            Author author1 = new Author("Lev", "Tolstoj");
+            Book book1 = new Book("226611188", "Anna", author1);
+            RB.AddBook(book1);
+
+            List<Book> books = RB.FindBooks( "War and Peace");
+
+            books.Count.ShouldBe(1);
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]: It's good that you have this test case, but it'd be covered if you fix the comment for `FindBookNameBook`. there's no right answer here, you just have to somehow maintain a balance between "one test case for each value" rule and having too huge amount of separate tests.

File: ClassLibrary/ClassLibrary/Reader.cs [hunk 18]
```diff
@@ -1,11 +1,73 @@
...
+        }
+
+        public Reader(String NameReader, String LastNameReader)
+        {
+
+            if (NameReader.IsNullOrWhiteSpace())
+            {
+                throw new ArgumentNullException("Name is empty");
+            }
+            else if (LastNameReader.IsNullOrWhiteSpace())
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Reader.cs]: usually we don't put `else` when there's `throw` or `return` in the `if` statement

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            {
+                request = (request == null) ?
+                    Books.Where(book => book.AuthorBook.LastName == LastNameAutor).ToList() :
+                    request.Where(book => book.AuthorBook.LastName == LastNameAutor).ToList();
+            }
+
+            return request.ToList();
+        }
+
+        //What AV1135?
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs]: Book is not a string, collection or task, so no problem here.

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+            else
+            {
+                Books.Add(book);
+            }
+        }
+
+        // what should we return ?
+        // what AV1130? I return "list" (For convenience), but use IEnumerable inside 
+        // And how should I convert the list (just below)? 
+        public List<Book> FindBooks(String NameBook="",String NameAutor="",String LastNameAutor="")
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs:38]: There are two problems with returning a reference type:
1. calling code can modify the returned collection, so if that's an internal collection and not a copy - the internal state of the called code will change, that's extremely error-prone (AV1130 is about that);
2. called code can modify the collection after returning it, so calling code might get different results for subsequent reads - that's also extremely error-prone;

To fix those:
1. return type should not allow modification (for example, `IEnumerable<T>` instead of `List<T>`)
2. return value should be a copy (ideally deep copy) of the internal object (for example, `return Books.ToList()` will make a copy)

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+            if (NameBook.IsNullOrWhiteSpace() &&
+                NameAutor.IsNullOrWhiteSpace() &&
+                LastNameAutor.IsNullOrWhiteSpace())
+            {
+                return Books;//??? convert IEnumerable->list?
+            }
+            ;
+            if (!NameBook.IsNullOrWhiteSpace())
+            {
+                request= Books.Where(book => book.NameBook == NameBook).ToList();
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs]: This is a very good approach! The problem is you're not following it :) 
I'll put it simple, you should read more about IEnumerable and enumerators to get a complete picture.

`IEnumerable` here is a list of commands that should be applied to a collection. All those commands are applied when you call `ToList()`. 

So to follow this approach of stacking commands you should only call `ToList()` once, in the end.

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+            }
+
+            if (!NameAutor.IsNullOrWhiteSpace())
+            {
+                request =(request == null)?
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs]: there's no need in any type of conditions inside this `if` statement. please do some thinking and make this whole method much simpler.

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            IEnumerable < Book > request = null;
+            if (NameBook.IsNullOrWhiteSpace() &&
+                NameAutor.IsNullOrWhiteSpace() &&
+                LastNameAutor.IsNullOrWhiteSpace())
+            {
+                return Books;//??? convert IEnumerable->list?
+            }
+            ;
+            if (!NameBook.IsNullOrWhiteSpace())
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs]: Negations worsen code readability. `IsNull` is also pretty negative, so we have double negation here, which makes the code impossible to just **read** - you have to **calculate** to understand it.
Please add an extension that will solve this problem.

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace ClassLibrary
+{
+    public class RepositoryBooks
+    {
+        // looks like with "DepartmentReaders", you can write for one occasion...  
+
+        //for comfortable I do it open, but I break AV1014 (An object should not expose any other classes it depends on for Library)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs:11]: but it's private, AV1014 is not violated here.

File: ClassLibrary/ClassLibrary/Library.cs [hunk 25]
```diff
@@ -6,59 +6,26 @@ namespace ClassLibrary
...
     {
-        public List<Book> Books
-        {
-            get;
-            private set;
-        }
-            = new List<Book>();
+        //You can see just below, I open method (I think this is inconvenient ), but I have private inside and It can be open RepositoryBooks here?
+        //look RepositoryBook - I break AV1014 (An object should not expose any other classes it depends on)
+        private RepositoryBooks repositoryBooks = new RepositoryBooks();
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs:9]: as long as you don't expose `RepositoryBooks` it's good.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs (hunk 15 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]: не понял, что ты хотел... 
и это DTO? https://docs.microsoft.com/en-us/aspnet/web-api/overview/data/using-web-api-with-entity-framework/part-5
и не совсем понимаю если они возвращают только часть запроса, а мне надо полностью книги этого автора

Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]: ммм.. да, DTO некорректный термин. я имел в виду тип, объединяющий все параметры поиска.

Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]:
```
class BooksFilter
{
  public string BookName {get; set;}
  public string AuthorName {get; set;}
  public string AuthorLastName {get; set;}
}
```

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs (hunk 16 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs:10]: согласен метод мне не нравиться, разбить на 2 метода один с двумя параметрами?

Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs:10]: а что это даст? тестирвать-то все равно надо все случаи. ты всё реализовал правильно, просто такую логику принципиально сложно тестить. но нужно.

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs (hunk 15 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs]: не совсем понятно, что с этой сущностью (классом) делать. 
Если в TestCase? - то необходимости объединять я не вижу, так как всеровно тебе придеться этот объект конструировать при каждом варианте запросаэто == проще сразу передавать три параметра...

File: ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs (hunk 16 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestRepositoryBooks.cs:10]: это даст уменьшение вариантов, но усложнения кода (в плане, что больше методов, неудобства поиска и т.д)
если два метода по 1-2 агрумента:
1 - 2 варианта (положительный отрицательный)
2 - 2*4  варианта (+/-)
=итого 10 вариантов против 16

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs (hunk 22 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/RepositoryBooks.cs]: I replace, but I am not sure that better solve (the condition exists)

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs (hunk 23 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/RepositoryBooks.cs]: I replace, but I am not sure that better solve

File: ClassLibrary/ClassLibrary/RepositoryBooks.cs [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+                Books.Add(book);
+            }
+        }
+
+ 
+        public List<Book> FindBooks(String NameBook="",String NameAutor="",String LastNameAutor="")
+        {
+            IEnumerable < Book > request = Books;
+
+            if (NameBook.IsNullOrWhiteSpace()==false&&request?.Count()!=0)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/RepositoryBooks.cs:42]: 1. Create an extension for `string` that will be the opposite of `IsNullOrWhiteSpace()` and use it here.
2. you don't need to check for `request.Count()`. and it iterates through the collection (`request`) every time, so affects performance.

Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary/RepositoryBooks.cs:42]: 3. the new code looks excellent, right?

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        public void BorrowBook(Reader reader,Book book)
+        {
+            if (!CheckReader(reader))
+            {
+                throw new ArgumentNullException($"Null Argument {((reader==null)?"reader":"book")} in AddBook");
+            }
+
+            if (!CheckReader(reader))
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/DepartmentReaders.cs:59]: this `if` statement is the same as above, it will never be true.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]
